undefined reference to `google::protobuf::internal::Release_CompareAndSwap' on s390x/Linux

##### System information (version)
- OpenCV => 3.4.x
- Operating System / Platform => Linux s390x 64bit
- Compiler => gcc 7.3.1

##### Detailed description

On s390x 64bit architecture, compilation fails on Linux:
[ 72%] Linking CXX executable ../../bin/opencv_perf_dnn
cd /git/opencv/build/modules/dnn && /usr/bin/cmake -E cmake_link_script CMakeFiles/opencv_perf_dnn.dir/link.txt --verbose=1
/usr/bin/c++     -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security   -Winit-self -Wpointer-arith   -Wuninitialized -Winit-self -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fno-omit-frame-pointer -ffunction-sections -fdata-sections  -fvisibility=hidden -fvisibility-inlines-hidden -Wno-shadow -Wno-parentheses -Wno-maybe-uninitialized -Wno-sign-promo -Wno-missing-declarations -Wno-deprecated -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-unused-local-typedefs -Wno-sign-compare -Wno-sign-promo -Wno-undef -Wno-ignored-qualifiers -Wno-extra -Wno-unused-function -Wno-unused-const-variable -Wno-deprecated-declarations -Wno-unused-parameter -Wno-undef -Wno-ignored-qualifiers -Wno-enum-compare -Wno-deprecated-declarations -Wno-invalid-offsetof -O3 -DNDEBUG  -DNDEBUG    -Wl,--gc-sections   -rdynamic CMakeFiles/opencv_perf_dnn.dir/perf/perf_caffe.cpp.o CMakeFiles/opencv_perf_dnn.dir/perf/perf_convolution.cpp.o CMakeFiles/opencv_perf_dnn.dir/perf/perf_main.cpp.o CMakeFiles/opencv_perf_dnn.dir/perf/perf_net.cpp.o  -o ../../bin/opencv_perf_dnn -Wl,-rpath,/git/opencv/build/lib ../../lib/libopencv_ts.a ../../lib/libopencv_dnn.so.3.4.3 ../../lib/libopencv_highgui.so.3.4.3 -ldl -lm -lpthread -lrt ../../lib/libopencv_videoio.so.3.4.3 ../../lib/libopencv_imgcodecs.so.3.4.3 ../../lib/libopencv_imgproc.so.3.4.3 ../../lib/libopencv_core.so.3.4.3 
../../lib/libopencv_dnn.so.3.4.3: undefined reference to `google::protobuf::internal::Release_CompareAndSwap(long volatile*, long, long)'
collect2: error: ld returned 1 exit status
make[2]: *** [modules/dnn/CMakeFiles/opencv_perf_dnn.dir/build.make:180: bin/opencv_perf_dnn] Error 1
make[2]: Leaving directory '/git/opencv/build'
make[1]: *** [CMakeFiles/Makefile2:4136: modules/dnn/CMakeFiles/opencv_perf_dnn.dir/all] Error 2
make[1]: Leaving directory '/git/opencv/build'
make: *** [Makefile:163: all] Error 2

##### Steps to reproduce

Simply try to compile OpenCV on s390x/Linux 64bit following the guide.

##### Correction

After a quick search on Google, it appears that this issue was solved in official protobuf code:
  - Reported issue: https://github.com/protocolbuffers/protobuf/issues/3937
  - PR: https://github.com/protocolbuffers/protobuf/pull/3955

Multiple solutions are possible:
  - patch the embedded protobuf code
  - upgrade embedded protobuf to lastest version

If patching the code is the right solution I have the patch ready and tested. Let me know if you want it (or you can just use the PR from protobuf ;-)).

--- a/3rdparty/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h
+++ b/3rdparty/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h
@@ -146,6 +146,14 @@ inline Atomic64 NoBarrier_Load(volatile const Atomic64* ptr) {
   return __atomic_load_n(ptr, __ATOMIC_RELAXED);
 }
 
+inline Atomic64 Release_CompareAndSwap(volatile Atomic64* ptr,
+                                       Atomic64 old_value,
+                                       Atomic64 new_value) {
+  __atomic_compare_exchange_n(ptr, &old_value, new_value, false,
+                              __ATOMIC_RELEASE, __ATOMIC_ACQUIRE);
+  return old_value;
+}
+
 #endif // defined(__LP64__)
 
 }  // namespace internal

